userdocs/diagnostics/associated-type-requirements.md

```diff
@@ -0,0 +1,18 @@
+# Using Protocols with Self or Associated Type Requirements
+
+A protocol with `Self` or associated type requirements may only be used to constrain the type of a generic parameter. This is because if a protocol named `Foo` has a requirement which references either `Self` or an associated type, then the type spelled `Foo` does not conform to the protocol `Foo` because it is unable to satisfy that requirement. As a result, `Foo` cannot be used as the type of a variable, argument, return value, etc. It may be used to constrain the type of a generic parameter `Bar` which represents a specific type.
```

Comment:
I think this section it should mention that in positions like argument, result use of names always refers to a type, in case of `Foo` it would be a value of protocol type, then continue with `and a value of type spelled as `Foo` doesn't conform to the protocol `Foo` because ...`. WDYT?